libraries/AP_GPS/AP_GPS.cpp

```diff
@@ -523,51 +619,96 @@ AP_GPS::update_instance(uint8_t instance)
 void
 AP_GPS::update(void)
 {
-    for (uint8_t i=0; i<GPS_MAX_INSTANCES; i++) {
+    for (uint8_t i=0; i<GPS_MAX_RECEIVERS; i++) {
         update_instance(i);
     }
 
-    // work out which GPS is the primary, and how many sensors we have
-    for (uint8_t i=0; i<GPS_MAX_INSTANCES; i++) {
+    // calculate number of instances
+    for (uint8_t i=0; i<GPS_MAX_RECEIVERS; i++) {
         if (state[i].status != NO_GPS) {
             num_instances = i+1;
         }
-        if (_auto_switch) {            
-            if (i == primary_instance) {
-                continue;
-            }
-            if (state[i].status > state[primary_instance].status) {
-                // we have a higher status lock, change GPS
-                primary_instance = i;
-                continue;
-            }
-
-            bool another_gps_has_1_or_more_sats = (state[i].num_sats >= state[primary_instance].num_sats + 1);
-
-            if (state[i].status == state[primary_instance].status && another_gps_has_1_or_more_sats) {
+    }
 
-                uint32_t now = AP_HAL::millis();
-                bool another_gps_has_2_or_more_sats = (state[i].num_sats >= state[primary_instance].num_sats + 2);
+    // if blending is requested, attempt to calculate weighting for each GPS
+    if (_auto_switch == 2) {
+        _output_is_blended = calc_blend_weights();
+        // adjust blend health counter
+        if (!_output_is_blended) {
+            _blend_health_counter = MIN(_blend_health_counter+BLEND_COUNTER_FAILURE_INCREMENT, 100);
+        } else if (_blend_health_counter > 0) {
+            _blend_health_counter--;
+        }
+    } else {
+        _output_is_blended = false;
+    }
 
-                if ( (another_gps_has_1_or_more_sats && (now - _last_instance_swap_ms) >= 20000) ||
-                     (another_gps_has_2_or_more_sats && (now - _last_instance_swap_ms) >= 5000 ) ) {
-                // this GPS has more satellites than the
-                // current primary, switch primary. Once we switch we will
-                // then tend to stick to the new GPS as primary. We don't
-                // want to switch too often as it will look like a
-                // position shift to the controllers.
-                primary_instance = i;
-                _last_instance_swap_ms = now;
+    if (_output_is_blended) {
+        // Use the weighting to calculate blended GPS states
+        calc_blended_state();
+        // set primary to the virtual instance
+        primary_instance = GPS_MAX_RECEIVERS;
+    } else {
+        // use switch logic to find best GPS
+        uint32_t now = AP_HAL::millis();
+        if (_auto_switch >= 1) {
+            // handling switching away from blended GPS
+            if (primary_instance == GPS_MAX_RECEIVERS) {
+                primary_instance = 0;
+                for (uint8_t i=1; i<GPS_MAX_RECEIVERS; i++) {
```

Comment:
I actually tried to do it as you suggested but I found the result was actually more complex and involved non-trivial changes to the existing logic and I didn't want to introduce extra uncertainty that the existing behaviour remains as it was previously.
Without going into all the details the issue is that in the existing switching logic it sets the _last_instance_swap_ms as soon as it finds a new GPS to use.  that's fine if there's only two but if you have 3 GPSs it means that it could initially choose a GPS that has a lower satellite count than the other.  Of course there are ways to work around this but like I said, it involves surgery on the existing logic.